modules/redis/scanner.go

```diff
@@ -0,0 +1,254 @@
+// distinct. The probe sends a sequence of commands and checks that the response
+// is well-formed redis data, which should be possible whatever the
+// configuration.
+package redis
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zmap/zgrab2"
+)
+
+// Flags contains redis-specific command-line flags.
+type Flags struct {
+	zgrab2.BaseFlags
+
+	// TODO: Take a JSON/YAML file with a list of custom commands to execute?
+	// TODO: Take a JSON/YAML file with mappings for command names?
+	AuthCommand string `long:"auth-command" default:"AUTH" description:"Override the command used to authenticate. Ignored if no password is set."`
+	Password    string `long:"password" description:"Set a password to use to authenticate to the server. WARNING: This is sent in the clear."`
+	DoInline    bool   `long:"inline" description:"Send commands using the inline syntax"`
+	Verbose     bool   `long:"verbose" description:"More verbose logging, include debug fields in the scan results"`
+}
+
+// Module implements the zgrab2.Module interface
+type Module struct {
+}
+
+// Scanner implements the zgrab2.Scanner interface
+type Scanner struct {
+	config *Flags
+}
+
+// Scan holds the state for the scan of an individual target
+type Scan struct {
+	scanner *Scanner
+	result  *Result
+	target  *zgrab2.ScanTarget
+	conn    *Connection
+}
+
+// Result is the struct that is returned by the scan.
+// If authentication is required, most responses can have the value
+// "(error: NOAUTH Authentication required.)"
+type Result struct {
+	// Commands is the list of commands actually sent to the server, serialized
+	// in inline format (e.g. COMMAND arg1 "arg 2" arg3)
+	Commands []string `json:"commands,omitempty" zgrab:"debug"`
+
+	// CommandOutput is the output returned by the server for each command sent;
+	// the index in CommandOutput matches the index in Commands.
+	CommandOutput [][]byte `json:"command_output,omitempty" zgrab:"debug"`
+
+	// PingResponse is the response from the server, should be the simple string
+	// "PONG".
+	// NOTE: This is invoked *before* calling AUTH, so this may return an auth
+	// required error even if --password is provided.
+	PingResponse string `json:"ping_response,omitempty"`
+
+	// InfoResponse is the response from the INFO command: "Lines can contain a
+	// section name (starting with a # character) or a property. All the
+	// properties are in the form of field:value terminated by \r\n."
+	InfoResponse string `json:"info_response,omitempty"`
+
+	// QuitResponse is the response from the QUIT command -- should be the
+	// simple string "OK" even when authentication is required, unless the
+	// QUIT command was renamed.
+	QuitResponse string `json:"quit_response,omitempty"`
+
+	// NonexistentResponse is the response to the non-existent command; even if
+	// auth is required, this may give a different error than existing commands.
+	NonexistentResponse string `json:"nonexistent_response,omitempty"`
```

Comment:
Is NONEXISTANT an actual command? Or are we testing the error response of a nonexistant command?

Reply:
The latter -- though even if it does for some reason exist, it is still the response to the command "NONEXISTENT" in the same sense that QuitResponse is the response to the command "QUIT".